File: clients/golang/examples/chunked_upload.go
package main

import (
	"context"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"time"
)

func createLargeTestFile(filePath string, sizeMB float64) error {
	fmt.Printf("Creating %.1fMB test file...\n", sizeMB)
	
	chunkSize := 1024 * 1024 // 1MB chunks
	totalChunks := int(sizeMB)
	
	file, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer file.Close()
	
	for i := 0; i < totalChunks; i++ {
		// Create varied content to make it more realistic
		content := fmt.Sprintf("Chunk %d/%d - %s\n", 
			i+1, totalChunks, 
			string(make([]byte, chunkSize-50)))
		
		if len(content) > chunkSize {
			content = content[:chunkSize]
		}
		
		_, err := file.WriteString(content)
		if err != nil {
			return err
		}
	}
	
	return nil
}

type DetailedProgress struct {
	startTime time.Time
	lastTime  time.Time
	lastBytes int64
}

func NewDetailedProgress() *DetailedProgress {
	now := time.Now()
	return &DetailedProgress{
		startTime: now,
		lastTime:  now,
	}
}

func (p *DetailedProgress) Callback(current, total int64) {
	now := time.Now()
	elapsed := now.Sub(p.startTime)
	
	// Calculate progress
	percent := float64(current) / float64(total) * 100
	
	// Calculate speed
	bytesSinceLast := current - p.lastBytes
	timeSinceLast := now.Sub(p.lastTime)
	
	var speedMBps float64
	if timeSinceLast.Seconds() > 0 {
		speedBps := float64(bytesSinceLast) / timeSinceLast.Seconds()
		speedMBps = speedBps / (1024 * 1024)
	}
	
	// Estimate remaining time
	var etaMinutes float64
	if current > 0 && elapsed.Seconds() > 0 {
		rate := float64(current) / elapsed.Seconds()
		remainingBytes := total - current
		etaSeconds := float64(remainingBytes) / rate
		etaMinutes = etaSeconds / 60
	}
	
	fmt.Printf("Progress: %5.1f%% | %s/%s | Speed: %5.1f MB/s | ETA: %4.1fm\n",
		percent,
		formatBytes(current),
		formatBytes(total),
		speedMBps,
		etaMinutes)
	
	p.lastTime = now
	p.lastBytes = current
}

func simulateResume(client *TUSClient, uploadURL, filePath string) {
	fmt.Println("\n🔄 Demonstrating upload resume capability...")
	
	ctx := context.Background()
	
	// Get current upload status
	uploadInfo, err := client.GetUploadInfo(ctx, uploadURL)
	if err != nil {
		fmt.Printf("❌ Failed to get upload info: %v\n", err)
		return
	}
	
	if offset, ok := uploadInfo["offset"].(int64); ok {
		if total, ok := uploadInfo["length"].(int64); ok {
			fmt.Printf("Current offset: %s (%.1f%%)\n", 
				formatBytes(offset), 
				float64(offset)/float64(total)*100)
			
			if offset >= total {
				fmt.Println("Upload already completed!")
				return
			}
		}
	}
	
	fmt.Println("This demonstrates how you would resume an interrupted upload...")
	fmt.Println("(In a real scenario, this would continue from where it left off)")
}

func main() {
	fmt.Println("TUS Chunked Upload Example (Golang)")
	fmt.Println("===================================")
	
	// Configuration
	endpoint := os.Getenv("TUS_ENDPOINT")
	if endpoint == "" {
		endpoint = "http://localhost:1080/files/"
	}
	
	fmt.Printf("Endpoint: %s\n", endpoint)
	fmt.Println()
	
	// Create TUS client
	client, err := NewTUSClient(endpoint)
	if err != nil {
		fmt.Printf("❌ Failed to create TUS client: %v\n", err)
		os.Exit(1)
	}
	
	// Create larger test file
	testFile := filepath.Join(os.TempDir(), "chunked_upload_test.txt")
	fileSizeMB := 5.0 // 5MB test file
	
	err = createLargeTestFile(testFile, fileSizeMB)
	if err != nil {
		fmt.Printf("❌ Failed to create test file: %v\n", err)
		os.Exit(1)
	}
	defer os.Remove(testFile) // Cleanup
	
	// Upload file with detailed progress tracking
	fmt.Printf("\nStarting chunked upload of %.1fMB file...\n", fileSizeMB)
	fmt.Printf("Using %s chunks\n", formatBytes(512*1024))
	
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
	defer cancel()
	
	progress := NewDetailedProgress()
	
	result, err := client.UploadFile(ctx, testFile, UploadOptions{
		ChunkSize: 512 * 1024, // 512KB chunks (smaller for demo)
		Metadata: map[string]string{
			"description": fmt.Sprintf("Chunked upload example - %.1fMB file", fileSizeMB),
			"upload_type": "chunked",
			"client":      "Golang TUS Client",
			"chunk_size":  "524288", // 512KB
		},
		ProgressCallback: progress.Callback,
	})
	
	if err != nil {
		fmt.Printf("❌ Upload failed: %v\n", err)
		
		// Show any partial upload info if available
		fmt.Println("\nAttempting to get partial upload information...")
		// This would require the upload URL from a partially successful upload
		
		os.Exit(1)
	}
	
	// Success
	fmt.Println("\n✅ Chunked upload completed successfully!")
	fmt.Printf("Upload URL: %s\n", result.URL)
	fmt.Printf("File size: %s (%.1f MB)\n", formatBytes(result.Size), float64(result.Size)/(1024*1024))
	fmt.Printf("Duration: %s\n", formatDuration(result.Duration))
	fmt.Printf("Checksum: %s\n", result.Checksum)
	
	// Calculate average speed
	if result.Duration.Seconds() > 0 {
		speedMBps := float64(result.Size) / result.Duration.Seconds() / (1024 * 1024)
		fmt.Printf("Average speed: %.1f MB/s\n", speedMBps)
	}
	
	fmt.Println("\nMetadata:")
	for key, value := range result.Metadata {
		fmt.Printf("  %s: %s\n", key, value)
	}
	
	// Get detailed upload information
	fmt.Println("\n📊 Upload Information:")
	uploadInfo, err := client.GetUploadInfo(ctx, result.URL)
	if err != nil {
		fmt.Printf("⚠️ Failed to get upload info: %v\n", err)
	} else {
		for key, value := range uploadInfo {
			fmt.Printf("  %s: %v\n", key, value)
		}
	}
	
	// Demonstrate resume capability
	fmt.Println("\n🎭 Resume Demonstration:")
	simulateResume(client, result.URL, testFile)
	
	fmt.Println("\n🎉 Chunked upload example completed successfully!")
}

File: clients/python/examples/chunked_upload.py
#!/usr/bin/env python3
"""
Chunked Upload Example

This example demonstrates uploading large files in chunks with resume capability.
"""

import os
import sys
import time
from pathlib import Path

# Add parent directory to path for imports
sys.path.append(str(Path(__file__).parent.parent))

from upload_client import TUSClient, TUSUploadError

def create_large_test_file(file_path: Path, size_mb: float = 10.0) -> Path:
    """Create a larger test file for chunked upload demonstration."""
    print(f"Creating {size_mb}MB test file...")
    
    chunk_size = 1024 * 1024  # 1MB chunks
    total_chunks = int(size_mb)
    
    with open(file_path, 'wb') as f:
        for i in range(total_chunks):
            # Create varied content to make it more realistic
            content = f"Chunk {i+1}/{total_chunks} - " + "x" * (chunk_size - 50) + "\n"
            f.write(content.encode()[:chunk_size])
    
    return file_path

class ProgressTracker:
    """Helper class to track upload progress with detailed statistics."""
    
    def __init__(self, total_size: int):
        self.total_size = total_size
        self.start_time = time.time()
        self.last_update = self.start_time
        self.last_bytes = 0
        
    def __call__(self, current: int, total: int):
        now = time.time()
        elapsed = now - self.start_time
        
        # Calculate progress
        percent = (current / total) * 100
        
        # Calculate speed
        bytes_since_last = current - self.last_bytes
        time_since_last = now - self.last_update
        
        if time_since_last > 0:
            speed_bps = bytes_since_last / time_since_last
            speed_mbps = speed_bps / (1024 * 1024)
        else:
            speed_mbps = 0
        
        # Estimate remaining time
        if current > 0 and elapsed > 0:
            rate = current / elapsed
            remaining_bytes = total - current
            eta_seconds = remaining_bytes / rate if rate > 0 else 0
            eta_minutes = eta_seconds / 60
        else:
            eta_minutes = 0
        
        print(f"Progress: {percent:5.1f}% | "
              f"{current:,}/{total:,} bytes | "
              f"Speed: {speed_mbps:5.1f} MB/s | "
              f"ETA: {eta_minutes:4.1f}m")
        
        self.last_update = now
        self.last_bytes = current

def simulate_interruption(client: TUSClient, upload_url: str, file_path: Path):
    """Simulate upload interruption and resume."""
    print(f"\n🔄 Simulating upload interruption and resume...")
    
    # Get current upload status
    upload_info = client.get_upload_info(upload_url)
    current_offset = upload_info['offset']
    total_size = upload_info['length']
    
    print(f"Current offset: {current_offset:,} bytes ({current_offset/total_size*100:.1f}%)")
    
    if current_offset >= total_size:
        print("Upload already completed!")
        return
    
    print("Resuming upload...")
    progress_tracker = ProgressTracker(total_size)
    
    result = client.resume_upload(
        upload_url=upload_url,
        file_path=file_path,
        progress_callback=progress_tracker
    )
    
    print(f"✅ Resume completed! Final URL: {result.url}")

def main():
    """Main example function."""
    # Configuration
    TUS_ENDPOINT = os.environ.get('TUS_ENDPOINT', 'http://localhost:1080/files/')
    
    print(f"TUS Chunked Upload Example")
    print(f"Endpoint: {TUS_ENDPOINT}")
    print("-" * 60)
    
    # Initialize TUS client with smaller chunks for demonstration
    client = TUSClient(
        endpoint=TUS_ENDPOINT,
        chunk_size=512 * 1024,  # 512KB chunks (smaller for demo)
        timeout=60,
        max_retries=5,
        retry_delay=2.0
    )
    
    # Create larger test file
    test_file = Path("/tmp/chunked_upload_test.txt")
    file_size_mb = 5.0  # 5MB test file
    create_large_test_file(test_file, size_mb=file_size_mb)
    
    upload_url = None
    
    try:
        # Upload file with detailed progress tracking
        print(f"\nStarting chunked upload of {file_size_mb}MB file...")
        print(f"Using {client.chunk_size:,} byte chunks")
        
        progress_tracker = ProgressTracker(int(file_size_mb * 1024 * 1024))
        
        result = client.upload(
            file_path=test_file,
            metadata={
                'filename': test_file.name,
                'description': f'Chunked upload example - {file_size_mb}MB file',
                'upload_type': 'chunked',
                'client': 'Python TUS Client',
                'chunk_size': str(client.chunk_size)
            },
            progress_callback=progress_tracker
        )
        
        upload_url = result.url
        
        print(f"\n✅ Chunked upload completed successfully!")
        print(f"Upload URL: {result.url}")
        print(f"File size: {result.size:,} bytes ({result.size/(1024*1024):.1f} MB)")
        print(f"SHA256 checksum: {result.checksum}")
        
        # Demonstrate upload info retrieval
        print(f"\n📊 Upload Information:")
        upload_info = client.get_upload_info(result.url)
        for key, value in upload_info.items():
            print(f"  {key}: {value}")
        
        # Demonstrate resume capability (simulate interruption)
        print(f"\n🎭 Resume Demonstration:")
        print("This would be useful if the upload was actually interrupted...")
        simulate_interruption(client, upload_url, test_file)
        
    except TUSUploadError as e:
        print(f"❌ Upload failed: {e}")
        
        # If we have an upload URL, try to get info about the failed upload
        if upload_url:
            try:
                upload_info = client.get_upload_info(upload_url)
                print(f"Failed upload info: {upload_info}")
            except:
                pass
        
        return 1
        
    except KeyboardInterrupt:
        print(f"\n⚠️ Upload interrupted by user")
        
        # If we have an upload URL, show current status
        if upload_url:
            try:
                upload_info = client.get_upload_info(upload_url)
                print(f"Interrupted upload info: {upload_info}")
                print(f"You could resume this upload later using: client.resume_upload('{upload_url}', '{test_file}')")
            except:
                pass
        
        return 1
        
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        return 1
        
    finally:
        # Clean up test file
        if test_file.exists():
            print(f"\nCleaning up test file: {test_file}")
            test_file.unlink()
    
    return 0

if __name__ == "__main__":
    sys.exit(main())